File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    cnt = 0
    big_bro = Counter(topping)
    little_bro = set()
    
    for i in topping:
        big_bro[i] -= 1
        if big_bro[i] == 0:
            del big_bro[i]
        
        little_bro.add(i)
        if len(big_bro) == len(little_bro):
            cnt += 1
    
    return cnt

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    
    Map<Integer, Integer> toppings;
    Set<Integer> bigBro = new HashSet<>();
    Set<Integer> littleBro = new HashSet<>();
    
    public int solution(int[] topping) {
        cntToppings(topping);
        int answer = 0;
        
        for (int i = 0; i < topping.length; i++) {
            int num = topping[i];  // 현재 토핑 
            
            littleBro.add(num);
            if (toppings.get(num) == 1) bigBro.remove(num);
            toppings.put(num, toppings.get(num) - 1);
            
            if (bigBro.size() == littleBro.size()) answer++;
        }
        
        return answer;
    }
    
    private void cntToppings(int[] topping) {
        toppings = new HashMap<>();

        for (int t : topping) {
            toppings.put(t, toppings.getOrDefault(t, 0) + 1);
            bigBro.add(t);
        }
    }
}